lucetc/src/compiler/cpu_features.rs

```diff
@@ -0,0 +1,159 @@
+use crate::error::{LucetcError, LucetcErrorKind};
+use cranelift_codegen::{isa, settings::Configurable};
+use failure::{format_err, ResultExt};
+use target_lexicon::Triple;
+
+/// x86 CPU families used as shorthand for different CPU feature configurations.
+///
+/// Matches the definitions from `cranelift-codegen`'s x86 settings definition.
+#[derive(Debug, Clone, Copy)]
+pub enum TargetCpu {
+    Baseline,
+    Nehalem,
+    Haswell,
+    Broadwell,
+    Skylake,
+    Cannonlake,
+    Icelake,
+    Znver1,
+}
+
+/// A manual specification of the CPU features to use during codegen.
+#[derive(Debug, Clone, Copy)]
+pub struct SpecificFeatures {
+    pub has_sse3: bool,
+    pub has_ssse3: bool,
+    pub has_sse41: bool,
+    pub has_sse42: bool,
+    pub has_popcnt: bool,
+    pub has_avx: bool,
+    pub has_bmi1: bool,
+    pub has_bmi2: bool,
+    pub has_lzcnt: bool,
+}
+
+impl SpecificFeatures {
+    /// Return a `CpuFeatures` with all optional features disabled.
+    pub fn all_disabled() -> Self {
+        Self {
+            has_sse3: false,
+            has_ssse3: false,
+            has_sse41: false,
+            has_sse42: false,
+            has_popcnt: false,
+            has_avx: false,
+            has_bmi1: false,
+            has_bmi2: false,
+            has_lzcnt: false,
+        }
+    }
+}
+
+impl From<TargetCpu> for SpecificFeatures {
+    fn from(cpu: TargetCpu) -> Self {
+        use TargetCpu::*;
+        match cpu {
+            Baseline => Self::all_disabled(),
+            Nehalem => Self {
+                has_sse3: true,
+                has_ssse3: true,
+                has_sse41: true,
+                has_sse42: true,
+                has_popcnt: true,
+                ..Baseline.into()
+            },
+            Haswell => Self {
+                // Note: this is not part of the Cranelift profile for Haswell, which only uses
+                // CPUID detection to enable AVX. If we want to bypass CPUID when compiling, we need
+                // to set it manually, and Haswell is the first of the CPUs with profiles to have
+                // AVX.
+                has_avx: true,
+                has_bmi1: true,
+                has_bmi2: true,
+                has_lzcnt: true,
+                ..Nehalem.into()
+            },
+            Broadwell => Haswell.into(),
+            Skylake => Broadwell.into(),
+            Cannonlake => Skylake.into(),
+            Icelake => Cannonlake.into(),
+            Znver1 => Self {
+                has_sse3: true,
+                has_ssse3: true,
+                has_sse41: true,
+                has_sse42: true,
+                has_popcnt: true,
+                // Note: similarly to the Haswell AVX flag, we don't want to rely on CPUID detection
+                // here, but Ryzen does support AVX.
+                has_avx: true,
+                has_bmi1: true,
+                has_bmi2: true,
```

Comment:
It might be worth a note here since we're talking about specific architectures - we technically don't have a profile that properly describes Bulldozer (which has BMI1 and not BMI2). I'm not sure if Cranelift even uses those instructions, and the likelihood of it affecting people seems low, but worth at least being aware of.

Reply:
Cranelift at least detects them; I'm not sure if it does anything interesting with them, though. My thinking is that someone wanting to target profile we don't have baked-in can always specify the feature set manually. It's not too hard to add new ones though, so I may transcribe a bit more of the gcc manpage...